main.nf

```diff
@@ -140,392 +165,666 @@ process fastqc {
+    memory { 3.GB * task.attempt }
+    time { 8.h * task.attempt }
+    errorStrategy { task.exitStatus == 143 ? 'retry' : 'terminate' }
+    maxRetries 3
+    maxErrors '-1'
 
-    publishDir "$results_path/trim_galore"
+    publishDir "${params.outdir}/trim_galore", mode: 'copy'
 
     input:
-    file(reads) from trimgalore_input
+    set val(name), file(reads:'*') from trimgalore_input
 
     output:
-    file '*fq.gz' into trimmed_reads, trimmed_reads_miRdeep2
-    file '*trimming_report.txt'
-    
+    file '*.gz' into trimmed_reads_bowtie, trimmed_reads_bowtie2
+    file '*trimming_report.txt' into trimgalore_results
+
     script:
-    /* Trim Galore should automatically detect the small RNA adapter sequence */
-    """
-    trim_galore --gzip --fastqc_args "-q" ${reads}
-    """
+    /* I'm using the latest version of trimgalore downloaded from github. The version on uppmax has a bug when handling smallRNA-seq data. */
+
+    single = reads instanceof Path
+
+    if(single) {
+      """
+      perl /home/chuanw/trimgalore/trim_galore.pl --small_rna --gzip $reads
+      """
+    } else {
+      """
+      perl /home/chuanw/trimgalore/trim_galore.pl --paired --small_rna --gzip $reads
+      """
+    }
 }
 
 
 /*
- * STEP 3 - Bowtie against miRBase mature RNA
+ * STEP 3.1 - Bowtie miRBase mature miRNA
  */
+
 process bowtie_miRBase_mature {
-    
+
+    tag "$name"
+
     module 'bioinfo-tools'
     module 'bowtie'
     module 'samtools'
-    
-    memory '2 GB'
-    time '6h'
+
+    cpus 2
+    memory { 16.GB * task.attempt }
+    time { 120.h * task.attempt }
+    errorStrategy { task.exitStatus == 143 ? 'retry' : 'terminate' }
+    maxRetries 3
+    maxErrors '-1'
+
+    publishDir "${params.outdir}/bowtie_miRBase_mature", mode: 'copy'
 
     input:
-    file(reads:'*') from trimmed_reads
-    
+    file(reads:'*') from trimmed_reads_bowtie
+    set val(name) from name_for_bowtie_miRBase_mature
+
     output:
-    file '*.bam' into miRBase_mature_bam
-    
+    file '*.mature.sam' into miRBase_mature_sam
+    file '*.mature_unmapped.fq.gz' into mature_unmapped_reads
+
     script:
-    """
-    f='$reads';f=(\$f);f=\${f[0]};f=\${f%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%_val_1};f=\${f%_trimmed};f=\${f%_1}
-    bowtie \
-        -p 2 \
-        -t \
-        -n 0 \
-        -l 15 \
-        -e 99999 \
-        -k 200 \
-        --best \
-        -S \
-        --chunkmbs 2048 \
-        $bowtie_miRBase_mature \
-        $reads \
-        | samtools view -bS > \${f}.aligned_miRBase_mature.bam
-    """
-}
+    /* Note! Only read 1 is used for alignment. */
+
+    single = reads instanceof Path
+
+    if(single) {
+      """
+      f='$reads';f=(\$f);input=\${f[0]};f=\${input%.gz};f=\${f%.fastq};f=\${f%.fq};f=\${f%.R1_val_1*};f=\${f%.R2_val_2*};f=\${f%_val_1};f=\${f%_val_2};f=\${f%_trimmed};f=\${f%_1}
+      prefix=\$f
+
+      bowtie $mature -q <(zcat \$input) -v 0 -k 10 -S \${prefix}.mature.sam --un \${prefix}.mature_unmapped.fq
```

Comment:
Also I'd prefer to pipe straight to samtools and save `.bam` files directly. Seems unnecessary to do that as a separate step and will use quite a bit more disk space.